### PR TITLE
Using StringBuilder instead of string

### PR DESCRIPTION
**Pull request by mayank408**
Fixes issue #725 

StringBuilder is added instead of String during concatenation.

### CODE REVIEW

**Comment by mayank408** [comment]
@chiragw15 @betterclever @amitiwary999 @chashmeetsingh Pls review.

**Comment by chiragw15** [comment]
https://github.com/fossasia/susi_android/blob/development/app/src/main/java/org/fossasia/susi/ai/activities/MainActivity.java#L284  
https://github.com/fossasia/susi_android/blob/development/app/src/main/java/org/fossasia/susi/ai/activities/SettingsActivity.java#L54
https://github.com/fossasia/susi_android/blob/development/app/src/main/java/org/fossasia/susi/ai/adapters/recycleradapters/ChatFeedRecyclerAdapter.java#L671
https://github.com/fossasia/susi_android/blob/development/app/src/main/java/org/fossasia/susi/ai/adapters/recycleradapters/ChatFeedRecyclerAdapter.java#L1006
https://github.com/fossasia/susi_android/blob/development/app/src/main/java/org/fossasia/susi/ai/adapters/recycleradapters/ChatFeedRecyclerAdapter.java#L1060
https://github.com/fossasia/susi_android/blob/development/app/src/main/java/org/fossasia/susi/ai/helper/MapHelper.java#L10
https://github.com/fossasia/susi_android/blob/development/app/src/main/java/org/fossasia/susi/ai/helper/MapHelper.java#L12
@mayank408 Here are some references. Change them here too. I may have missed some, so check once again.

**Comment by mayank408** [comment]
@chiragw15 There is no need to change every String, like the one which are inline. Pls review @iamareebjamal

**Comment by mayank408** [comment]
@chiragw15 It does make sense to change every string to stringBuilder.

**Comment by chiragw15** [comment]
Okay, I am not sure about this. @iamareebjamal can tell better.

File: app/src/main/java/org/fossasia/susi/ai/adapters/recycleradapters/ChatFeedRecyclerAdapter.java [hunk 1]
```diff
@@ -1072,20 +1072,20 @@ public void onClick(DialogInterface dialog, int which) {
...
                             ChatMessage message = getData().get(i);
                             if (message.getActionType()==null || message.getActionType().equals(Constant.ANSWER)) {
                                 Log.d(TAG, message.toString());
-                                shareText += "[" + message.getTimeStamp() + "]";
-                                shareText += " ";
-                                shareText += message.isMine() ? "Me: " : "Susi: ";
-                                shareText += message.getContent();
-                                shareText += "\n";
+                                shareText.append("[").append(message.getTimeStamp()).append("]");
+                                shareText.append(" ");
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/adapters/recycleradapters/ChatFeedRecyclerAdapter.java]: append is chainable. Chain it like you have done in previous line

**Comment by iamareebjamal** [comment]
@chiragw15 There is no need to do inline strings. In fact, string builder performs poorly than string concatenation for small additions like 2-3 strings

File: app/src/main/java/org/fossasia/susi/ai/adapters/recycleradapters/ChatFeedRecyclerAdapter.java (hunk 1 above)
Comment (mayank408) [inline app/src/main/java/org/fossasia/susi/ai/adapters/recycleradapters/ChatFeedRecyclerAdapter.java]: Changes done @iamareebjamal Please review and approve.

File: app/src/main/java/org/fossasia/susi/ai/adapters/recycleradapters/ChatFeedRecyclerAdapter.java [hunk 2]
```diff
@@ -1051,20 +1048,17 @@ public void onClick(DialogInterface dialog, int which) {
...
-                                shareText += message.isMine() ? "Me: " : "Susi: ";
-                                shareText += message.getContent();
-                                shareText += "\n";
+                                shareText.append("[").append(message.getTimeStamp()).append("]").append(" ");
+                                shareText.append(message.isMine() ? "Me: " : "Susi: ").append(message.getContent()).append("\n");
                             }
                         }
-                        shareText = shareText.substring(0, shareText.length() - 1);
-                        shareMessage(shareText);
+                        String shareText2 = shareText.substring(0, shareText.length() - 1);
```
Comment (betterclever) [inline app/src/main/java/org/fossasia/susi/ai/adapters/recycleradapters/ChatFeedRecyclerAdapter.java]: Why are you extracting string in this way? You could use shareText.toString()

File: app/src/main/java/org/fossasia/susi/ai/adapters/recycleradapters/ChatFeedRecyclerAdapter.java [hunk 3]
```diff
@@ -1014,20 +1014,17 @@ public void onClick(DialogInterface dialog, int which) {
...
-                                copyText += message.isMine() ? "Me: " : "Susi: ";
-                                copyText += message.getContent();
-                                copyText += "\n";
+                                copyText.append("[").append(message.getTimeStamp()).append("]").append(" ");
+                                copyText.append(message.isMine() ? "Me: " : "Susi: ").append(message.getContent()).append("\n");
                             }
                         }
-                        copyText = copyText.substring(0, copyText.length() - 1);
-                        setClipboard(copyText);
+                        String copyText2 = copyText.substring(0, copyText.length() - 1);
```
Comment (betterclever) [inline app/src/main/java/org/fossasia/susi/ai/adapters/recycleradapters/ChatFeedRecyclerAdapter.java]: Here too.

**Comment by iamareebjamal** [comment]
@mayank408 There are other places you can use StringBuilder too. Please search for the places where there are more than 3 concatenations. Thanks :smile:

File: app/src/main/java/org/fossasia/susi/ai/adapters/recycleradapters/ChatFeedRecyclerAdapter.java [hunk 4]
```diff
@@ -1014,20 +1015,16 @@ public void onClick(DialogInterface dialog, int which) {
                         }
                         setClipboard(copyText);
                     } else {
-                        String copyText = "";
+                        StringBuilder copyText = new StringBuilder("");
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/adapters/recycleradapters/ChatFeedRecyclerAdapter.java]: No need to add a double quote as an argument in StringBuilder

File: app/src/main/java/org/fossasia/susi/ai/helper/MapHelper.java [hunk 5]
```diff
@@ -6,10 +6,10 @@
 import org.fossasia.susi.ai.model.MapData;
 
 public class MapHelper {
-    private static final String BASE_URL = "http://staticmap.openstreetmap.de/staticmap.php?";
-    private static final String URL_SCHEME = BASE_URL + "center=%f,%f&zoom=%f&size=%sx%s";
-    private static final String MAP_URL = "https://www.openstreetmap.org/#map=";
-    private static final String MAP_URL_SCHEME = MAP_URL + "%f/%f/%f";
+    private static final StringBuilder BASE_URL = new StringBuilder("http://staticmap.openstreetmap.de/staticmap.php?");
+    private static final StringBuilder URL_SCHEME = BASE_URL.append("center=%f,%f&zoom=%f&size=%sx%s");
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/helper/MapHelper.java]: These were single concats. Please only use StringBuilder when more than 3 concats are there

**Comment by betterclever** [comment]
@mayank408 There are many places in current diff, in which StringBuilder is used even when there are just concatenation 1-2 times. Is it necessary in all those places? It might be improving performance , if it really is, but decreases code legibility to a great extent there.
So, as @iamareebjamal pointed out, we must do it in places with more than 3 concatenations and not all the places perhaps. 
@iamareebjamal Correct me if I am wrong.

File: app/src/main/java/org/fossasia/susi/ai/helper/MapHelper.java (hunk 5 above)
Comment (mayank408) [inline app/src/main/java/org/fossasia/susi/ai/helper/MapHelper.java]: Sorry but I could not found where there is 3 concats. Thanks

File: app/src/main/java/org/fossasia/susi/ai/adapters/recycleradapters/ChatFeedRecyclerAdapter.java (hunk 4 above)
Comment (mayank408) [inline app/src/main/java/org/fossasia/susi/ai/adapters/recycleradapters/ChatFeedRecyclerAdapter.java]: Okay

**Comment by iamareebjamal** [comment]
@betterclever Correct. Somehow my review did not get posted on github where I said the same. Besides concatenating 2 strings is better than StringBuilder performance wise too. Only use StringBuilder when concatenating more than 3 strings. @chiragw15 has already given many sources where StringBuilder is needed

**Comment by mayank408** [comment]
@iamareebjamal I checked out the resources given by @chiragw15 . The thing we cant use string builder in URI's and in other I have used it.

File: app/src/main/java/org/fossasia/susi/ai/helper/MapHelper.java (hunk 5 above)
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/helper/MapHelper.java]: Then please fall back to String concat in these places. Thanks

**Comment by mayank408** [comment]
@chiragw15 @iamareebjamal I could not find any use of StringBuilder other than in copy and share method. There is no need to forcefully implement something. So I think we should restrict ourself to that only . Thanks.

**Comment by betterclever** [comment]
@iamareebjamal I see there are some more places, but I am just curious that given that StringBuilder optimizations are introduced in Java 6 compiler which we are targeting, based on our minimum API level like this answers point out.
https://stackoverflow.com/questions/11942368/why-use-stringbuilder-explicitly-if-the-compiler-converts-string-concatenation-t
Do we need it in places without a loop?

**Comment by iamareebjamal** [comment]
Android implementations are different from Oracle JDK implementation. And it is good practice because nevertheless, it stops you from recreating immutable string objects at each concatenation

**Comment by mayank408** [comment]
@iamareebjamal All changes done.

File: app/src/main/java/org/fossasia/susi/ai/adapters/recycleradapters/ChatFeedRecyclerAdapter.java (hunk 4 above)
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/adapters/recycleradapters/ChatFeedRecyclerAdapter.java]: This remains to change. Please remove the constructor parameter

**Comment by mayank408** [comment]
@iamareebjamal done

**Comment by the-dagger** [comment]
@mayank408 Please squash your commits.

**Comment by mayank408** [comment]
@chiragw15 @amitiwary999  Pls approve this.

**Pull request merged**